Changes for upcoming remote observing API

joshjriley opened the pull request.

joshjriley committed:
keck_vnc_config.yaml

Added api_key config.  Updated instructions for firewall config.

joshjriley committed:
Updated to use API nominally while still retaining old method of using firewall config and SSH calls as a backup.  Still some work to do.

joshjriley committed:
Added 'i' option to menu to View extra connection info for viewing VNC password and Zoom info

joshjriley committed:
Requery API for sessions list if called from menu

joshjriley committed:
Added vncports command line option.

joshjriley committed:
Added vncserver cmd line option.

joshjriley committed:
Update README.md

joshjriley committed:
Update README.md

joshjriley committed:
updated api port to "telnetport"

joshjriley commented:
Do we want to get rid of instructions to generate a helpdesk ticket to "Notify Keck of your intent to connect remotely"?  Part of that is to send along all the info about their remote/home station.  If so, should we have them enter that info on the web with their SSH key upload?  Would require a little bit of dev time to expand the webform and figure out where in the DB to put that info.

joshjriley commented:
The test_all portion currently has API 'account' param hard-coded to 'kcwi1'.  The API requires an instrument account which we don't necessarily have when they run '--test'.  We could change the API to accept no account, but Jeff suggested instead we get rid of the connection test to all 10 vm-instrs and instead just have them test a real instr account during their test window.

joshjriley commented:
I think there is a prior bug when firewall password is entered incorrectly which results in a strange failure.  Code flounders around a bit doing other stuff after trying to exit app.  This pushes the warning that the "API key is not set" off the screen. 
 We might want to fix this b/c this might come up a lot when existing users don't follow instructions to add api_key and we change the fw pass on them.

joshwalawender commented:
I haven't dug in to the details of the code yet (soon I hope), but I think for the failure you mention in the most recent comment, we should add a warning for anyone not using the API.  

Another way to distinguish old setups (i.e. current users) who haven't adopted the new API yet would be to deprecate the current username for the firewall password and use something else in the future.  Then we just check if they're using the old username in the config file and error out for anyone still using that firewall user name.

joshwalawender commented on keck_vnc_launcher.py:
```diff
@@ -18,16 +19,18 @@
 
-__version__ = '1.2.4'
+## Module vars
+__version__ = '1.3.0'
```
I suggest going to v2.0.0, this is a big change.

joshwalawender commented on keck_vnc_launcher.py:
```diff
@@ -91,6 +94,10 @@ def create_parser():
         help="Path to local configuration file.")
+    parser.add_argument("-vncserver", type=str,
+        help="Name of VNC server to connect to.  Takes precedence over all.")
+    parser.add_argument( '-vncports', nargs='+', type=str,
```
I suggest double dashes to follow the conventions already in use, so `--vncserver` and `--vncports`

joshjriley replied:
whoops, just an oversight.  will fix.

joshwalawender commented on keck_vnc_launcher.py:
```diff
@@ -447,6 +458,16 @@ def start(self):
+            self.get_api_data(self.api_key, self.args.account)
+            if self.api_data is None:
+                if self.firewall_defined is True:
+                    self.log.info('Firewall info detected in config.  Trying ate method.')
```
"ate method"?

joshjriley replied:
"alternate".  Must have made a mistake editing text.  will fix.

joshjriley committed:
fix firewall error typo

joshjriley committed:
fix config dashes

joshjriley committed:
bump version to 2.0.0

joshwalawender commented on keck_vnc_launcher.py:
```diff
@@ -1067,17 +1179,7 @@ def validate_ssh_key(self):
-
-        # Find a server in the servers_to_try list with is not an svncserverN
-        server = None
-        for s in self.servers_to_try:
```
Why is this block being removed?  I think it is still necessary (though only in the case where the API is not functioning of course).

joshjriley replied:
I can’t even remember editing that function, so I think it was a rogue edit that didn’t get caught in my testing b/c it still works.  I’ll put it back to what it was.

joshjriley committed:
remove erroneous hash value. Was a test and is not a valid hash value.

joshjriley committed:
Returning rogue delete of a few lines of code to original.

joshjriley committed:
Added proc.returncode to do_ssh_command.  Allow validate_ssh_check to only check return code is ok if we are using API

joshjriley commented:
mistakenly close pull request

joshwalawender commented on keck_vnc_launcher.py:
```diff
@@ -824,6 +863,79 @@ def get_vncviewer_properties(self):
+        #form API url and get data
+        url = f'{KRO_API}?key={self.api_key}'
+        if account is not False: url += f'&account={self.args.account}'
+        self.log.debug(f'Calling API with URL: {url}')
```
We should also log what responses we got from the API.  If we don't want to put some info in the log for security reasons, we should just have a log line which says that we got that info.

joshjriley replied:
ok done.  Logging API return code and if successfully parsed.  I'm thinking we shouldn't log the actual response b/c of all the sensitive info in it.  Easy to add if we change mind.

joshwalawender commented:
This is looking pretty good, but we haven't closed the loop on the VNC password.  We either need to print that to the screen so the user can type it in, or we need to decide to do away with it.

joshwalawender commented on keck_vnc_launcher.py:
```diff
@@ -1114,23 +1229,38 @@ def get_vnc_server(self, account, instrument):
-                break
+        #API Route
+        elif self.api_data:
+            self.log.info(f"Determining VNC server for '{self.args.account}' (via API)")
```
Need to log what server we got, similar to line 1262 below.

joshjriley replied:
ok done

joshjriley committed:
change default api err msg

joshjriley committed:
Show connection info by default

joshjriley committed:
Log API response code and log if successfully parsed it.

joshjriley committed:
Log vncserver value for either API or SSH route

joshjriley commented:
> This is looking pretty good, but we haven't closed the loop on the VNC password. We either need to print that to the screen so the user can type it in, or we need to decide to do away with it.

I added call to view_connection_info at the very end just before re-displaying menu.  Added  border around it to make it extra visible.

joshjriley committed:
Fixed incorrect exit procedure when firewall password is entered incorrectly.

joshjriley committed:
Removed test code

joshjriley committed:
Merge branch 'issue81' into api

joshwalawender committed:
Update readme for new instructions
